supervisor/supervisor.go

```diff
@@ -276,6 +291,21 @@ func (s *Supervisor) Start() error {
 		"cpus":        s.machine.Cpus,
 	}).Debug("containerd: supervisor running")
 	go func() {
+		if (len(restoreExitTasks) > 0) {
+			logrus.Infof("containerd: found %d exited processes after restart", len(restoreExitTasks))
+			//
+			// If the restore() function stored any ExitTask in the dedicated slice,
+			// then handle those tasks first. The purpose of the one second delay is
+			// to give dockerd a chance to establish its event stream connection to
+			// containerd. If the connection is established before the ExitTask are
+			// being handled, then dockerd can receive exit notifications directly
+			// (rather than having to replay the notifications from the event log).
+			//
+			time.Sleep(time.Second)
+			for _, e := range restoreExitTasks {
+				s.handleTask(e)
```

Comment:
just another question: should it be `s.SendTask(e)`?  Since we are dropping this part at 388?

Reply:
The crucial section of the patched code now looks like this.

```
    293         go func() {
    294                 if (len(restoreExitTasks) > 0) {
    295                         logrus.Infof("containerd: found %d exited processes after restart", len(restoreExitTasks))
     :
    304                         time.Sleep(time.Second)
    305                         for _, e := range restoreExitTasks {
    306                                 s.handleTask(e)
    307                         }
    308                 }
    309                 for i := range s.tasks {
    310                         s.handleTask(i)
    311                 }
    312         }()

    323 func (s *Supervisor) SendTask(evt Task) {
    324         select {
     :
    328         case s.tasks <- evt:
    329                 TasksCounter.Inc(1)
    330         }
    331 }

```
If SendTask() would be called at line 306 instead of handleTask(), the goroutine would get blocked at line 328 in SendTask() if len(restoreExitTasks) is greater than 2048 (defaultBufferSize) because the s.tasks queue would be filled entirely with ExitTask entries. In other words, containerd would still be prone to the problem that this pull request aims to address. The task handler loop at lines 309..311 is not running yet, so the s.tasks queue does not get drained. Therefore, handleTask() needs to be called directly.